jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java

```diff
@@ -695,6 +696,56 @@ public void run() {
         logger.info("Sessions found {}", sessionsInfo.stream().map(SessionInfo::getId).collect(Collectors.toList()));
         em.close();
     }
+
+    @Test
+    public void testPerCaseEnsureCleanupInDisposeEngine() throws InterruptedException {
+        Map<String, OrganizationalEntity> roleAssignments = new HashMap<>();
+        roleAssignments.put("owner", new UserImpl("john"));
+
+        Map<String, Object> data = new HashMap<>();
+        data.put("s", "description");
+        CaseFileInstance caseFile = caseService.newCaseFileInstance(deploymentUnit.getIdentifier(), USER_TASK_CASE_P_ID, data, roleAssignments);
+        CaseDefinition caseDef = caseRuntimeDataService.getCase(deploymentUnit.getIdentifier(), USER_TASK_CASE_P_ID);
+        assertNotNull(caseDef);
+        assertEquals(deploymentUnit.getIdentifier(), caseDef.getDeploymentId());
+        assertEquals(3, caseDef.getAdHocFragments().size());
+        String caseId = caseService.startCase(deploymentUnit.getIdentifier(), USER_TASK_CASE_P_ID, caseFile);
+        
+        Collection<ProcessInstanceDesc> desc = runtimeDataService.getProcessInstancesByDeploymentId(deploymentUnit.getIdentifier(), Arrays.asList(0,1,2,3,4), new QueryContext());
+        List<Long> ids = desc.stream().map(e -> e.getId()).collect(Collectors.toList());
+        Long pid = ids.get(0);
+
+        RuntimeManager manager = RuntimeManagerRegistry.get().getManager(deploymentUnit.getIdentifier());
+        RuntimeEngine engine = manager.getRuntimeEngine(ProcessInstanceIdContext.get(pid));
+        // no op so the runtime is not being init when lazy loading is used
+        manager.disposeRuntimeEngine(engine);
+        try {
+            Thread executor = new Thread(new Runnable () {
+                @Override
+                public void run() {
+                    RuntimeEngine localEngine = manager.getRuntimeEngine(ProcessInstanceIdContext.get(pid));
+                    manager.disposeRuntimeEngine(localEngine);
+                }
+                
+            });
+            executor.start();
+            executor.join();
```

Comment:
For making the test more robust (and make it fail if session is null, instead of waiting here indefinitely), _join()_ might add a timeout, and then check that the thread is not alive:
```suggestion
            executor.join(10000);
            assertFalse(executor.isAlive());
```

Reply:
done